[WIP] debug xml dsig

fisx opened the pull request:
These changes make XML signature validation more robust, and also more correct.

Tasks:

- [x] capture stdout, stderr noise from libxml2 and feed it into an error that is returned;
- [x] make sure that the signed subtrees are self-contained by applying an extra round of canonicalization;
- [x] fix test suite (currently: `"failed to canonicalize input: xmlReadMemory: illegal operation (Inappropriate ioctl for device)"`)

fisx committed:
Tweaks (#2)

* Add support for SHA256 signature digests.

* Improve errors.

* Use -XTypeApplications.

* rm trailing whitespace.

* Add test cases (signature verification).

* haddocks.

* introduce hpack (helped with fixing a ghci-vs.-libxml-2.0 issue).

* Compatibility with GHC-8.4 (#1)

fisx committed:
Tweaks (#3)

* Cleanup.

* Silence confused warning.

* More verbose sign verification errors; more exports.

fisx committed:
rm hsaml2.cabal in favor of package.yaml

stack got confused about stale hsaml2.cabal files from time to time.

fisx committed:
Cleanup

fisx committed:
Fix: remove impossible error case.

fisx committed:
Cleanup

fisx committed:
Cleanup

fisx committed:
Fix: do not lose namespace declarations when cutting out signed subtree.

fisx committed:
Fix: capture stdout, stderr and translate into locatable error.

fisx committed:
Cleanup

fisx committed:
Fixup 64b0562476f9c1277bbf246114c908665b0fc750

fisx committed:
Haddocks

fisx commented:
feel free to ignore; i'm not sure at all i'm not breaking anything here, but some of the changes are needed for this to work for us.  one day i hope to push everything we have upstream to you again into a sane whole.   :-)

fisx committed:
Review

fisx committed:
Review

fisx committed:
Cleanup (polish error messages some more)

fisx committed:
Cleanup

dylex commented:
I haven't looked much at any of this in a while, nor do I have a good way to test it anymore, but I'm glad if it's at least useful. Much of this looks good and some I'd probably have to spend more time with to understand. I'm certainly happy to take all the forward-compatibility updates and cleanups, but don't feel obligated to do more work to split things out.  In many ways this package seems to be more useful as a starting point for people than a library since everyone tends to have slightly different considerations in ways that have been hard to generalize.

fisx committed:
Fix: a better approach to getting a signed subtree.

fisx committed:
Fix: how xpath appearently works.

fisx committed:
Make libxml2 parser wrapper handle stderr noise and errors better.

fisx committed:
Cleanup

fisx committed:
Fix: test suite.

fisx committed:
Fix: remove broken test case.

I think this test had some issue with whitespace being changed
after signing, and canonicalization does not touch whitespace
that much.  I am removing it because it is very likely that this
is the issue, and very hard to robustly confirm it.

fisx committed:
Add test case: idempotency of canonicalization.

Commit:
Cleanup

Co-Authored-By: fisx <mf@zerobuzz.net>

fisx commented:
thanks for the feedback!  i only opened this by accident, but i'm glad it's now documented here some more that there is a branch.  :-)

i willl close this, feel free anybody to steal stuff from the https://github.com/wireapp/hsaml2 fork!
